toqito/state_opt/bell_inequality_max.py

```diff
@@ -0,0 +1,168 @@
+    :param a_val: The value of each measurement outcome for A
+    :param b_val: The value of each measurement outcome for B
+    :return: The upper bound for the maximum violation of the Bell inequality
+
+
+
+    """
+    m, _ = joint_coe.shape
+    oa = len(a_val)
+    ob = len(b_val)
+
+    # Ensure the input vectors are column vectors
+    a_val = a_val.reshape(-1, 1)
+    b_val = b_val.reshape(-1, 1)
+    a_coe = a_coe.reshape(-1, 1)
+    b_coe = b_coe.reshape(-1, 1)
+
+    # Check if vectors a_val and b_val have only two elements
+    if oa != 2 or ob != 2:
+        raise ValueError("This script is only capable of handling Bell inequalities with two outcomes.")
+
+    tot_dim = 2 ** (2 * m + 2)
+    obj_mat = np.zeros((tot_dim, tot_dim), dtype=float)
+
+    # Nested loops to compute the objective matrix
+    for a in range(2):  # a = 0 to 1
+        for b in range(2):  # b = 0 to 1
+            for x in range(1, m + 1):  # x = 1 to m (1-indexed in MATLAB, hence the range adjustment)
+                for y in range(1, m + 1):  # y = 1 to m
+                    b_coeff = (
+                        joint_coe[x - 1, y - 1] * a_val[a, 0] * b_val[b, 0]
+                    )  # Adjust index for 0-based Python indexing
+                    if y == 1:
+                        b_coeff += a_coe[x - 1, 0] * a_val[a, 0]  # Adjust for 0-based indexing
+                    if x == 1:
+                        b_coeff += b_coe[y - 1, 0] * b_val[b, 0]  # Adjust for 0-based indexing
+
+                    # Adding the result of the tensor product to the objective matrix
+                    obj_mat += b_coeff * np.kron(MN_matrix(m, a, x), MN_matrix(m, b, y))
+
+    # Symmetrize the matrix to avoid numerical issues
+    obj_mat = (obj_mat + obj_mat.T) / 2
+    aux_mat = np.array([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
+
+    # Construct the SDP problem
+    W = cp.Variable((2 ** (2 * m), 2 ** (2 * m)), symmetric=True)
+
+    M = swap(W, [2, m + 1], [2] * (2 * m))
+    X = swap(cp.kron(M, aux_mat), [m + 1, 2 * m + 1], [2] * (2 * m + 2))
+    Z = swap(X, [m + 2, 2 * m + 1], [2] * (2 * m + 2))
+
+    objective = cp.Maximize(cp.trace(Z @ obj_mat))
+
+    # Define the constraints
+    constraints = [cp.trace(W) == 1, W >> 0]
+
+    # Adding PPT constraints
+    for sz in range(1, m + 1):
+        # Generate all combinations of indices from 1 to 2*m-1 of size sz
+        for ppt_partition in combinations(range(1, 2 * m - 1), sz):
+            # Convert to 0-based indexing for Python
+            ppt_partition_updated = [x - 1 for x in ppt_partition]
+            # Partial transpose on the partition, ensuring it's positive semidefinite
+            pt_matrix = partial_transpose(W, ppt_partition_updated, [4] + [2] * (2 * (m - 1)))
+            constraints.append(pt_matrix >> 0)
+
+    # Solve the problem
+    prob = cp.Problem(objective, constraints)
+    prob.solve(solver="MOSEK", verbose=False)

```

Comment:
We shouldn't be using Mosek by default as this is a proprietary solver that users may not have access to. 

Instead, we can make the solver an optional argument parameter to the `bell_inequality_max` function. Something like `solver: str = "cvxopt"` should be included in the argument parameter listing.